### PR TITLE
Update decomposition patterns for value semantics

### PR DESCRIPTION
**Pull request by bmhowe23**
* Update QuakeOperator builders to automatically return wires when appropriate.
* Create helpers to allow common "create" functions for both reference semantics and value semantics.
* Update UnitaryBuilder to understand unwraps.
* Update decomposition pattern tests to additionally test value semantics.

### CODE REVIEW

**Review by boschmitt — COMMENTED** [review]
LGTM, I found the design of `QuakeOperatorCreator` quite nice.

File: lib/Optimizer/Transforms/DecompositionPatterns.cpp [hunk 1]
```diff
@@ -40,6 +41,140 @@ inline Value createDivF(Location loc, Value numerator, double denominator,
...
 
+inline bool isAllWires(quake::OperatorInterface op) {
+  for (auto c : op.getControls())
+    if (!c.getType().isa<quake::WireType>())
+      return false;
+  for (auto t : op.getTargets())
+    if (!t.getType().isa<quake::WireType>())
+      return false;
+  return true;
+}
```
[the remark this answers is not in the document]
Comment (bmhowe23) [reply lib/Optimizer/Transforms/DecompositionPatterns.cpp:52]: > Perhaps use `quake::isLinearValueForm` function.

Unless I'm mistaken, I think that will allow `!quake.control` types though, and I don't want to let those types through because they are not fed through to the outputs.

File: lib/Optimizer/Transforms/DecompositionPatterns.cpp [hunk 2]
```diff
@@ -114,24 +249,29 @@ struct HToPhasedRx : public OpRewritePattern<quake::HOp> {
                                 PatternRewriter &rewriter) const override {
     if (!op.getControls().empty())
       return failure();
-    if (!quake::isAllReferences(op))
+    const bool allWires = isAllWires(op);
+    if (!quake::isAllReferences(op) && !allWires)
```
[the remark this answers is not in the document]
Comment (bmhowe23) [reply lib/Optimizer/Transforms/DecompositionPatterns.cpp:253]: > I wonder if you still need this test. Quantum operation are either memory semantics (all references) or value semantics (all wires). A mix should be ill-formed. If you can handle both, there shouldn't be a need to test.

@schweitzpgi - is this check already part of the verifiers somewhere? If so, I may be able to remove it as @boschmitt is suggesting. (Though I still would want to exclude `!quake.control` types because, as I understand it, those are not linear values.)

File: lib/Optimizer/Transforms/DecompositionPatterns.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (boschmitt) [reply lib/Optimizer/Transforms/DecompositionPatterns.cpp:52]: Tbh, I'm not familiar with `quake.control`. However, `quake::isLinearValueForm` is what the mapping pass uses to sanity check the ability to map the circuit. Thus, if this function is not a good fit here, then it is likely it shouldn't be a good fit there.

[the remark this answers is not in the document]
Comment (bmhowe23) [reply lib/Optimizer/Transforms/DecompositionPatterns.cpp:52]: > Tbh, I'm not familiar with `quake.control`. However, `quake::isLinearValueForm` is what the mapping pass uses to sanity check the ability to map the circuit. Thus, if this function is not a good fit here, then it is likely it shouldn't be a good fit there.

As a quick test, I tried updating `isAllWires(op)` to simply call `quake::isLinearValueForm(op)`, and that caused the decomposition tests to fail. The reason is because `quake::isLinearValueForm(op)` prohibits both a) an operation's operands from originating from a `quake.unwrap` operation, and b) an operation's results from being sent to a `quake.wrap` operation. That may be a good restriction to enforce during mapping (because that could prevent the mapping pass from comprehending all the operations), but I see no reason to enforce that for this pass. Also, that condition is precisely what happens with the current tests, since the current tests contain `quake.ref` function parameters.

In other words, we are using the `memtoreg` pass to allow the current tests to perform double duty testing (both reference semantics and value semantics), and as an example, that transforms

FROM:
```mlir
func.func @test(%control: !quake.ref, %target: !quake.ref) {
  quake.h [%control] %target : (!quake.ref, !quake.ref) -> ()
  return
}
```
TO:
```mlir
  func.func @test(%arg0: !quake.ref, %arg1: !quake.ref) {
    %0 = quake.unwrap %arg1 : (!quake.ref) -> !quake.wire
    %1 = quake.unwrap %arg0 : (!quake.ref) -> !quake.wire
    %2:2 = quake.h [%1] %0 : (!quake.wire, !quake.wire) -> (!quake.wire, !quake.wire)
    quake.wrap %2#0 to %arg0 : !quake.wire, !quake.ref
    quake.wrap %2#1 to %arg1 : !quake.wire, !quake.ref
    return
  }
```
And `quake::isLinearValueForm(H_Operation)` return false, which is undesirable in my opinion.

If we want to prohibit nearby `quake.unwrap` and `quake.wrap` instructions like the mapping pass does, then I would need to write new tests for all of these decomposition passes, and I (currently) see no need to enforce this tighter restriction on the decomposition pass. What do you think?

File: lib/Optimizer/Transforms/DecompositionPatterns.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (schweitzpgi) [reply lib/Optimizer/Transforms/DecompositionPatterns.cpp:253]: We need to support intermediate forms and other types (such as `!quake.control`). I would recommend removing the "all this or that" test(s), and just dealing with the arguments in a type correct way.

File: lib/Optimizer/Transforms/DecompositionPatterns.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (schweitzpgi) [reply lib/Optimizer/Transforms/DecompositionPatterns.cpp:52]: I don't see how these type changes matter at all with the semantics of the decomposition. They're orthogonal.

What we have is a plurality of ways to express a quantum operation using distinctive types to express different constraints in the IR explicitly and implicitly. The constraints, in fact, do not change. The operation (gate) does not change either.

Decomposition is the replacement of a gate, $a$, with a composition of gates, $a'_0, a'_1, ..., a'_n$. As stated above, the constraints don't change, they are either made explicit or implicit. So the expansion from $a$ to $a'_0, a'_1, ... a'_n$ should simply forward the argument types as presented to $a$ through $a'_0, a'_1, ... a'_n$ to the same end result.

```mlir
   %a:j = quake.a %args ... : (!quake.types ...) -> (!quake.types ...)
```
becomes
```mlir
   %a'_0:j = quake.a'_0 %args ... : (!quake.types ...) -> (!quake.types ...)
   %a'_1:j = quake.a'_1 %args ... : (!quake.types ...) -> (!quake.types ...)
   ...
   %a'_n:j = quake.a'_n %args ... : (!quake.types ...) -> (!quake.types ...)
```

File: include/cudaq/Optimizer/Dialect/Quake/QuakeOps.td [hunk 3]
```diff
@@ -852,15 +852,25 @@ class QuakeOperator<string mnemonic, list<Trait> traits = [],
                    "mlir::ValueRange":$controls,
                    "mlir::ValueRange":$targets,
                    "mlir::DenseBoolArrayAttr":$negates), [{
-      return build($_builder, $_state, mlir::TypeRange{}, is_adj, parameters,
+      return build($_builder, $_state, static_cast<bool>(is_adj), parameters,
```
Comment (schweitzpgi) [inline include/cudaq/Optimizer/Dialect/Quake/QuakeOps.td:855]: Probably want to just add a new `bool` builder overload that's missing the return type.

File: lib/Optimizer/Transforms/DecompositionPatterns.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (bmhowe23) [reply lib/Optimizer/Transforms/DecompositionPatterns.cpp:52]: This documentation makes it sound like `!quake.control` is never returned as the result of an operation: https://github.com/NVIDIA/cuda-quantum/blob/c4001f66a8f90c0367126b0fdc5df59f24cd8cdf/include/cudaq/Optimizer/Dialect/Quake/QuakeTypes.td#L80.
That seems to conflict with `(!quake.types ...) -> (!quake.types ...)` because the output types do not necessarily have to match the input types, especially when `!quake.control` is used.

[the remark this answers is not in the document]
Comment (schweitzpgi) [reply lib/Optimizer/Transforms/DecompositionPatterns.cpp:52]: On the question of "linear types" and the `!quake.control` type, we may be able to fine tune it a bit, but not without really understanding everywhere the test is currently deployed. (In that regard, it's easier to craft a new test for precisely the constraints to be checked than to break other unrelated code in subtle ways.)

`1quake.control` values are SSA values in implementation. However, they are also constrained such that they should only be used within a linear type representation and require "bounding" operations to transform a linear type value to a `control` and convert it back. I believe that was precisely what the test was intended to test for.

File: include/cudaq/Optimizer/Dialect/Quake/QuakeOps.td (hunk 3 above)
Comment (bmhowe23) [reply include/cudaq/Optimizer/Dialect/Quake/QuakeOps.td:855]: > Probably want to just add a new `bool` builder overload that's missing the return type.

Is that what is in the next few lines?

File: lib/Optimizer/Transforms/DecompositionPatterns.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (bmhowe23) [reply lib/Optimizer/Transforms/DecompositionPatterns.cpp:253]: > We need to support intermediate forms and other types (such as `!quake.control`). I would recommend removing the "all this or that" test(s), and just dealing with the arguments in a type correct way.

I believe that means we will have to do `quake.from_ctrl` and `quake.to_ctrl` in cases shown below.

https://github.com/NVIDIA/cuda-quantum/blob/c4001f66a8f90c0367126b0fdc5df59f24cd8cdf/lib/Optimizer/Transforms/DecompositionPatterns.cpp#L487
https://github.com/NVIDIA/cuda-quantum/blob/c4001f66a8f90c0367126b0fdc5df59f24cd8cdf/lib/Optimizer/Transforms/DecompositionPatterns.cpp#L657
(Cases like the ones linked above perform gates _on the controls_, treating the controls _as targets_ for part of the decompositions.)

In other words, just because a qubit is used as a `quake.control` on the _input_ to the decomposition, doesn't mean that it can remain a `quake.control` _throughout_ the decomposition.

[the remark this answers is not in the document]
Comment (bmhowe23) [reply lib/Optimizer/Transforms/DecompositionPatterns.cpp:253]: The decomposition algorithm used in the "CCZToCX" pattern appears to be fundamentally incompatible with the notion of accepting `!quake.control` types because it modifies the state of the control qubit.
https://github.com/NVIDIA/cuda-quantum/blob/c4001f66a8f90c0367126b0fdc5df59f24cd8cdf/lib/Optimizer/Transforms/DecompositionPatterns.cpp#L602-L610

[the remark this answers is not in the document]
Comment (boschmitt) [reply lib/Optimizer/Transforms/DecompositionPatterns.cpp:253]: > The decomposition algorithm used in the "CCZToCX" pattern appears to be fundamentally incompatible with the notion of accepting `!quake.control` types because it modifies the state of the control qubit.

FYI, this pattern illustrates well one reason why we decided to go back on our initial design in which controls did not generate new values. We concluded it was a bad tradeoff.

One one hand, not generating new values for controls allowed us to easily identify equivalences like:
```
   (1)  (2)  (3)       (2)  (1)  (3)       (1)  (3)  (2)        
────●────●────●───   ───●────●────●───   ───●────●────●───
    │    │    │         │    │    │         │    │    │  
    │  ┌─┴─┐  │       ┌─┴─┐  │    │         │    │  ┌─┴─┐
────┼──┤ x ├──┼─── = ─┤ x ├──┼────┼─── = ───┼────┼──┤ x ├─
    │  └───┘  │       └───┘  │    │         │    │  └───┘
  ┌─┴─┐     ┌─┴─┐          ┌─┴─┐┌─┴─┐     ┌─┴─┐┌─┴─┐     
──┤ x ├─────┤ x ├─   ──────┤ x ├┤ x ├─   ─┤ x ├┤ x ├──────
  └───┘     └───┘          └───┘└───┘     └───┘└───┘     
```
Mitigating the problem of structural bias, and thus enhancing the effectiveness of some optimizations (e.g., identifying that (1) and (3) cancel out).

On the other hand, this idea made changing the IR more error prone in general. For example, this rewriting pattern changes controls to targets, thus all uses of the control value from the rewriting point onwards would need to change, but the preceding uses would need to stay the same.

File: lib/Optimizer/Transforms/DecompositionPatterns.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (bmhowe23) [reply lib/Optimizer/Transforms/DecompositionPatterns.cpp]: > Perhaps use `quake::isLinearValueForm` function.

I updated all decompositions to support mixed `quake.wire` / `quake.ref` semantics in 2056fd81c. In that commit, I also updated all decompositions to support mixed `quake.wire` / `quake.ref` / `quake.control` semantics where possible. The function that started this thread (`isAllWires()`) has now been removed, so I'm marking this as resolved. Please feel free to open another thread if we want to continue the discussion on a different topic.

**Review by schweitzpgi — APPROVED** [review]
Nice work, Ben. Thank you.

File: lib/Optimizer/Transforms/DecompositionPatterns.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (schweitzpgi) [reply lib/Optimizer/Transforms/DecompositionPatterns.cpp]: For posterity: These transitions from `!quake.control` to targets can be handled seamlessly with the use of `quake::FromControlOp` and `quake::ToControlOp`. For the above, after the seventh gate (X), we could insert a `quake::FromControlOp` and feed that wire into the eighth gate (X). Similarly for the top most wire in the figure. Finally, we'd add two `quake::ToControlOp` to convert the wire back to a control type for type correctness.

[the remark this answers is not in the document]
Comment (bmhowe23) [reply lib/Optimizer/Transforms/DecompositionPatterns.cpp]: Is it safe to assume that all uses a `!quake.control` have to exist in same block that generated it? In other words, is the following IR not allowed?

```mlir
func.func @test(%arg0: !quake.ref, %arg1: !quake.ref, %arg2: !quake.ref) {
  %0 = quake.unwrap %arg0 : (!quake.ref) -> !quake.wire
  %1 = quake.unwrap %arg1 : (!quake.ref) -> !quake.wire
  %2 = quake.unwrap %arg2 : (!quake.ref) -> !quake.wire
  %c0 = quake.to_ctrl %0 : (!quake.wire) -> !quake.control
  %cond = cc.undef i1
  %3:2 = cc.if(%cond) ((%a1 = %1, %a2 = %2)) -> (!quake.wire, !quake.wire) {
    %4:2 = quake.z [%c0, %a1] %a2 : (!quake.control, !quake.wire, !quake.wire) -> (!quake.wire, !quake.wire)
    cc.continue %4#0, %4#1 : !quake.wire, !quake.wire
  } else {
    cc.continue %a1, %a2 : !quake.wire, !quake.wire
  }
  %5 = quake.from_ctrl %c0 : (!quake.control) -> !quake.wire
  quake.wrap %5 to %arg0 : !quake.wire, !quake.ref
  quake.wrap %3#0 to %arg1 : !quake.wire, !quake.ref
  quake.wrap %3#1 to %arg2 : !quake.wire, !quake.ref
  return
}
```
The above IR currently passes all validation checks, but if we performed a decomposition on it, we would have nowhere to pass the new `!quake.control` value generated by the decomposition to (and the downstream code may attempt to keep using `%c0` or `%5` despite that no longer corresponding to the correct value anymore.

I was assuming that the above IR is valid, and it was therefore unsafe to try to do what you are suggesting, but maybe this should be flagged as invalid IR?

Or maybe we should use `isUsedOutsideOfBlock` to detect this condition and only `return failure()` in this condition?

[the remark this answers is not in the document]
Comment (schweitzpgi) [reply lib/Optimizer/Transforms/DecompositionPatterns.cpp]: This IR is fine.

It's a bit odd in that the region of the control is unnecessarily too large. So one strategy would be to thread wires into the if and move the to_control and from_control within the region they actually make sense.

Another valid transformation for this would have to take into account that `cc.if` has regions and definitions in those regions are not upwards exposed and realize the definition and at least 1 other use are not in the same region. So we could do something like:

```mlir
  %c0 = to_ctrl
  %if:3 = cc.if ...  {
    ... %c0 ....
    ...
    %new_wire = from_ctrl %c0
    ....
    %new_control = to_ctrl %new_wire'
    continue %new_control, ...
  } else {
    continue %c0, ...
  }
  ...
  %wire = from_control %if#0
```

So, yeah, it's still possible, but we have to maintain a valid SSA representation on the control typed values.

[the remark this answers is not in the document]
Comment (schweitzpgi) [reply lib/Optimizer/Transforms/DecompositionPatterns.cpp]:
```mlir
module {
  func.func @test(%arg0: !quake.ref, %arg1: !quake.ref, %arg2: !quake.ref) {
    %0 = quake.unwrap %arg0 : (!quake.ref) -> !quake.wire
    %1 = quake.unwrap %arg1 : (!quake.ref) -> !quake.wire
    %2 = quake.unwrap %arg2 : (!quake.ref) -> !quake.wire
    %3 = quake.to_ctrl %0 : (!quake.wire) -> !quake.control
    %4 = cc.undef i1
    %5:3 = cc.if(%4) ((%arg3 = %1, %arg4 = %2)) -> (!quake.control, !quake.wire, !quake.wire) {
      %7:2 = quake.x [%arg3] %arg4 : (!quake.wire, !quake.wire) -> (!quake.wire, !quake.wire)
      %8 = quake.t<adj> %7#1 : (!quake.wire) -> !quake.wire
      %9 = quake.x [%3] %8 : (!quake.control, !quake.wire) -> !quake.wire
      %10 = quake.t %9 : (!quake.wire) -> !quake.wire
      %11:2 = quake.x [%7#0] %10 : (!quake.wire, !quake.wire) -> (!quake.wire, !quake.wire)
      %12 = quake.t<adj> %11#1 : (!quake.wire) -> !quake.wire
      %13 = quake.x [%3] %12 : (!quake.control, !quake.wire) -> !quake.wire
      %14 = quake.x [%3] %11#0 : (!quake.control, !quake.wire) -> !quake.wire
      %15 = quake.t<adj> %14 : (!quake.wire) -> !quake.wire
      %16 = quake.x [%3] %15 : (!quake.control, !quake.wire) -> !quake.wire
      %17 = quake.t %16 : (!quake.wire) -> !quake.wire
      %18 = quake.t %13 : (!quake.wire) -> !quake.wire
      %19 = quake.from_ctrl %3 : (!quake.control) -> !quake.wire
      %20 = quake.t %19 : (!quake.wire) -> !quake.wire
      %21 = quake.to_ctrl %20 : (!quake.wire) -> !quake.control
      cc.continue %21, %17, %18 : !quake.control, !quake.wire, !quake.wire
    } else {
      cc.continue %3, %arg3, %arg4 : !quake.control, !quake.wire, !quake.wire
    }
    %6 = quake.from_ctrl %5#0 : (!quake.control) -> !quake.wire
    quake.wrap %6 to %arg0 : !quake.wire, !quake.ref
    quake.wrap %5#1 to %arg1 : !quake.wire, !quake.ref
    quake.wrap %5#2 to %arg2 : !quake.wire, !quake.ref
    return
  }
}
```

[the remark this answers is not in the document]
Comment (bmhowe23) [reply lib/Optimizer/Transforms/DecompositionPatterns.cpp]: OK, thanks. I think I recall a prior discussion about not threading `quake.control`s through `cc.if`s, but if so, that was probably describing the _current_ state of things and not saying that they _couldn't_ be threaded through them. That makes sense now. Thanks.

[the remark this answers is not in the document]
Comment (schweitzpgi) [reply lib/Optimizer/Transforms/DecompositionPatterns.cpp]: The code posted above is legal and can be run through `cudaq-opt`.

I think the confusion may be that values of type `!quake.control` are SSA. Values of type `!quake.wire` are SSI. In the context of a `cc.if` this distinction matters.

[the remark this answers is not in the document]
Comment (bmhowe23) [reply lib/Optimizer/Transforms/DecompositionPatterns.cpp]: Understood. Thanks.

FWIW, both of those Quake snippets in this thread don't seem to make it through `cudaq-opt -regtomem`. I'm not sure if we should create a separate issue for that, or if that is expected behavior? (The first one fails an `assert` check, and the second one doesn't perform any IR transformation.)
